src/Authentication/Authentication.Core/Authenticator.cs

```diff
@@ -77,36 +106,84 @@ public static async Task<IAuthContext> AuthenticateAsync(IAuthContext authContex
 
                 JwtHelpers.DecodeJWT(httpRequestMessage.Headers.Authorization?.Parameter, account, ref authContext);
-                return authContext;
+                return (authContext, new AuthError(AuthErrorType.None, null));
             }
             catch (AuthenticationException authEx)
             {
-                if ((authEx.InnerException is TaskCanceledException) && cancellationToken.IsCancellationRequested)
+                //Interactive Authentication Failure: Could Not Open Browser, fallback to DeviceAuth
+                if (IsUnableToOpenWebPageError(authEx))
+                {
+                    authContext.UseDeviceAuth = true;
+                    //ReAuthenticate using DeviceCode as fallback.
+                    var (retryAuthContext, retryAuthError) = await AuthenticateAsync(authContext, forceRefresh, cancellationToken);
+                    //Indicate that this was a Fallback
+                    retryAuthError = new AuthError(AuthErrorType.FallBack, retryAuthError.Exception ?? authEx);
+                    return (retryAuthContext, retryAuthError);
+                }
+                // DeviceCode Authentication Failure: Timeout
+                if (authEx.InnerException is TaskCanceledException && cancellationToken.IsCancellationRequested)
                 {
                     // DeviceCodeTimeout
-                    throw new Exception(string.Format(
+                    var deviceCode = new Exception(string.Format(
                             CultureInfo.CurrentCulture,
                             ErrorConstants.Message.DeviceCodeTimeout,
                             Constants.MaxDeviceCodeTimeOut));
+                    return (authContext, new AuthError(AuthErrorType.DeviceCodeFailure, deviceCode.InnerException ?? deviceCode));
                 }
-                else
-                {
-                    throw authEx.InnerException ?? authEx;
-                }
+                //Something Unknown Went Wrong
+                return (authContext, new AuthError(AuthErrorType.Unknown, authEx.InnerException ?? authEx));
             }
             catch (Exception ex)
             {
-                throw ex.InnerException ?? ex;
+                return (authContext, new AuthError(AuthErrorType.Unknown, ex.InnerException ?? ex));
             }
         }
 
         /// <summary>
-        /// Signs out of the provided <see cref="IAuthContext"/>.
+        ///  Gets a static instance of IAuthenticationProvider when the client app hasn't changed.
         /// </summary>
-        /// <param name="authContext">The <see cref="IAuthContext"/> to sign-out from.</param>
-        public static void LogOut(IAuthContext authContext)
+        /// <param name="authContext"></param>
+        /// <returns></returns>
+        private static (IAuthenticationProvider authProvider, IClientApplicationBase clientApplication, AuthProviderType authProviderType) GetClientApplication(IAuthContext authContext)
```

Comment:
This is not needed if `fallBackWarning` is made nullable in `AuthenticateAsync`. See my suggestion above.